File: 08-composite/08-composite.go
package main

import "fmt"

type Item interface {
	show()
}

type Product struct {
	Name string
}

func (p *Product) show() {
	fmt.Println(" -", p.Name)
}

type Box struct {
	Name     string
	Children []Item
}

func (b *Box) show() {
	fmt.Println(b.Name)
	for i := 0; i < len(b.Children); i++ {
		b.Children[i].show()
	}
}

func (b *Box) add(item Item) {
	b.Children = append(b.Children, item)
}

func main() {
	p1 := &Product{Name: "Shirt"}
	p2 := &Product{Name: "Suit"}
	p3 := &Product{Name: "Tie"}

	premium := &Box{Name: "Premium"}
	premium.add(p1)
	premium.add(p2)
	premium.add(p3)

	premium.show()
}

File: 08-composite/08-composite.js
class Item
{
    show() {
        throw new Error("not implemented");
    }
}

class Product extends Item {
    constructor(name) {
        super();
        this.name = name;
    }

    show() {
        return " - " + this.name;
    }
}

class Box extends Item {
    constructor(name) {
        super();
        this.name = name;
        this.children = [];
    }

    show() {
        console.log(this.name);
        this.children.forEach(c => console.log(c.show()));
    }

    add(item) {
        this.children.push(item);
    }
}

const p1 = new Product("phone");
const p2 = new Product("phone holder");
const p3 = new Product("charger");

const box = new Box("phone premium");

box.add(p1);
box.add(p2);
box.add(p3);

box.show();
